### PR TITLE
[Bugfix][Whisper] Apply timestamp decoding rules

### LINKED ISSUE #55281: [Bug]: Whisper segment timestamps collapse under `prompt` — the reference timestamp decoding rules are not applied
### Your current environment

- vLLM `v0.28.0` (official image `vllm/vllm-openai:v0.28.0`, plus `librosa soundfile av`)
- Model: `openai/whisper-large-v3-turbo`, `--enforce-eager --max-num-seqs 4`
- Endpoint: `POST /v1/audio/transcriptions`, `response_format=verbose_json`
- **The fix from #53207 (`<|startofprev|>`) was applied for every measurement below.** Without it the `prompt` field is broken for an unrelated reason and the numbers here cannot be reproduced.

### 🐛 Describe the bug

**Passing `prompt` to a `verbose_json` transcription collapses the segmentation.** The text survives; the timestamps do not. A 30 s clip that yields 9 segments without a prompt yields 1 segment with one, spanning the whole window.

This is not the `<|prev|>` bug (#53206) — that is fixed in these runs.

#### Measurements

30 s Mandarin public-service announcement, `language=zh`, `temperature=0`, deterministic across repeats. Prompt is a list of domain terms (the usual "hotwords" use case), lengthened word by word:

| `prompt` | prompt tokens | segments | chars | last segment ends |
|---|---|---|---|---|
| *(none)* | 0 | **9** | 165 | 30.4 |
| 4 words | 14 | 8 | 123 | 29.8 |
| 8 words | 29 | **2** | 140 | 30.4 |
| 12 words | 44 | **1** | 132 | 29.8 |
| 16 words | 56 | **2** | 149 | 30.4 |
| 20 words | 76 | 1 | 15 | *(output degenerated to `ご視聴ありがとうございました。`)* |
| 24 words | 87 | **0** | 0 | — |
| 30 words | 101 | **0** | 0 | — |

The prompt never approaches the decoder budget (101 tokens out of `max_model_len=448`), and the transcript itself stays complete up to 16 words — only the segmentation degrades.

#### The timestamps are not being emitted at all

`verbose_json` returns each segment's `tokens`. Timestamp tokens for this model start at `<|0.00|>` = **50365**. Counting them inside the returned segments:

```
no prompt   : 9 segments, 0 timestamp tokens inside any segment  (all at the boundaries, as expected)
12-word prompt : 1 segment, 139 tokens, 0 timestamp tokens inside
```

So the model transcribes the full 30 s but emits **no intermediate timestamps whatsoever**. Nothing is being lost by the parser — the information never exists.

#### The same weights do not behave this way under the reference implementation

`faster-whisper` 1.2.1 (CTranslate2 4.8.2), same checkpoint, same audio, same word lists.
vLLM ran fp16 on GPU, faster-whisper int8 on CPU — that difference shows up in the
transcribed text, not in whether timestamps are emitted:

| `prompt` | vLLM segments | faster-whisper segments |
|---|---|---|
| *(none)* | 9 | 8 |
| 4 words | 8 | 7 |
| 8 words | **2** | 6 |
| 12 words | **1** | 6 |
| 16 words | **2** | 7 |
| 20 words | **1** | 7 |
| 24 words | **0** | 3 |
| 30 words | **0** | 6 |

### Where it comes from

vLLM decodes Whisper as a generic encoder-decoder LM. The reference implementation does not: it constrains the logits at every step with `ApplyTimestampRules` (`whisper/decoding.py`), which

- suppresses `<|notimestamps|>` when timestamps are requested,
- forces timestamps to appear **in pairs** and to increase monotonically,
- requires the sequence to begin with a timestamp,
- bounds the first timestamp by `max_initial_timestamp`.

CTranslate2 implements the same rules inside its Whisper-specific `generate()`, which is why `faster-whisper` can expose `max_initial_timestamp` as an option at all.

vLLM has none of this. Instead, `_preprocess_verbose_prompt` in `vllm/entrypoints/speech_to_text/base/serving.py` string-replaces `<|notimestamps|>` with `<|0.00|>` in the decoder prefix:

```python
dec_prompt["prompt"] = dec_prompt["prompt"].replace("<|notimestamps|>", "<|0.00|>")
```

This is a stand-in for the missing rules: it guarantees the output *starts* as if in timestamp mode, and nothing after that keeps it there. A prompt — plain prose carrying no timestamp tokens — biases the decoder toward continuous text, and with no constraint to emit paired timestamps it simply stops emitting them.

### The obvious fix does not work — I tried it

Removing the replacement (so the model generates its own first timestamp, as the reference does) makes things worse, not better:

| `prompt` | segments before | segments after |
|---|---|---|
| *(none)* | 9 | **0** |
| 4 words | 8 | 8 |
| 8 words | 2 | **9** |
| 12 words | 1 | 1 |
| 16 words | 2 | 1 |

Logging the first generated token explains it:

```
no prompt      -> 50364  <|notimestamps|>   (the model opts out of timestamps entirely)
4-word prompt  -> 50365  <|0.00|>
12-word prompt -> 50365  <|0.00|>           (in timestamp mode, still only 1 segment)
```

Without `ApplyTimestampRules` suppressing it, `<|notimestamps|>` is simply a legal continuation and the model takes it. And the 12-word row shows that even entering timestamp mode is not enough: with no pairing or monotonicity constraint, the decoder can run to the end of the window without emitting another timestamp.

So the prefix is not the cause, and the prefix hack cannot be removed on its own.

### Suggested fix

Implement Whisper's timestamp rules as a logits processor applied to transcription requests when segment timestamps are requested — the four constraints listed above. That would let `_preprocess_verbose_prompt` drop the string replacement, and would make segment timestamps hold up under `prompt`.

This also matters for #53144: advancing the cursor to the end of the last segment relies on the timestamps being emitted correctly in the first place, which today is not guaranteed.

### Related

- #53206 / #53207 — `<|prev|>` prompt token; fixed in all measurements above
- #53144 — audio dropped when the decoder stops early; complementary, and depends on this
- #47664 — word-level timestamps
- #25750 — Whisper feature tracking

### Before submitting a new issue...

- [x] Made sure I searched the documentation and existing issues.

### PR DESCRIPTION
**Pull request by CharlesXu-HQ**
## Purpose

Fixes #55281.

Whisper timestamp requests did not apply the decoding rules required by the model. This change adds a native per-request timestamp logits processor for timestamp/text alternation, monotonic timestamps, bounded initial timestamps, and timestamp probability forcing. It also propagates the request rules through standard and beam-search generation, makes verbose JSON strip the no-timestamps token instead of forcing a zero timestamp, and preserves the replacement request's processor state when a persistent batch slot is recycled.

A duplicate search immediately before submission found no open or closed PR referencing #55281 or implementing this fix.

Implementation was assisted by OpenAI GPT-5. The author is responsible for reviewing and validating the change before it is marked ready for merge.

## Test Plan

- Cover initial timestamp bounds, timestamp/text alternation, monotonic timestamps, and timestamp mass forcing.
- Cover batched processor state movement and beam-search prompts with appended tokens.
- Cover removal and replacement at the same persistent batch index.
- Cover speech-to-text serving parameter propagation.
- Run the integrated timestamp test suite and Ruff checks.

The exact Whisper model checkpoint was not available on the target host, and outbound Hugging Face downloads were unavailable, so an end-to-end model evaluation was not run. The unit and integration tests exercise the processor rules and serving wiring directly.

## Test Result

- Targeted tests: `9 passed, 16 warnings in 2.44s`
- `ruff check`: all checks passed
- `ruff format --check`: 9 files already formatted

---

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, including the issue it resolves.
- [x] The test plan, including the test commands and coverage goals.
- [x] The test results, including targeted test and Ruff results.
- [x] Documentation impact considered; no documentation update is required for this bug fix.

</details>

**BEFORE SUBMITTING, PLEASE READ <https://docs.vllm.ai/en/latest/contributing>**

### CODE REVIEW

File: vllm/v1/sample/logits_processor/whisper.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        ) is None:
+            return None
+        return _WhisperTimestampRequest(
+            rules=rules,
+            prompt_token_ids=prompt_token_ids or (),
+            output_token_ids=output_token_ids,
+        )
+
+    def update_state(self, batch_update: BatchUpdate | None) -> None:
+        process_dict_updates(self.requests, batch_update, self._new_request)
```
[the remark this answers is not in the document]
Comment (CharlesXu-HQ) [reply vllm/v1/sample/logits_processor/whisper.py:93]: Fixed in d99abdf. process_dict_updates now follows the documented removed -> added -> moved order, and test_rules_survive_batch_slot_replacement covers a same-index replacement. The targeted suite passes with 9 tests.
